numpyro/distributions/batch_util.py

```diff
@@ -523,7 +524,7 @@ def _default_promote_batch_shape(d: Distribution):
     return new_self
 
 
-@promote_batch_shape.register
+@promote_batch_shape.register(Independent)
```

Comment:
This will disable behavior for ExpandedDistribution. You might need to have a separate register for Inpedendent (reuse _promote_batch_shape_expanded as in my comment). Anyway, could you add description for why this is needed. The default behavior does not work for Independent?

Reply:
Here's an example that reproduces the error I got:

```python
import numpyro
import jax.numpy as jnp
import jax.random as random
import numpy as np
import numpyro.distributions as dist
from numpyro.handlers import mask
from numpyro.infer import SVI, Trace_ELBO
from optax import adamw
from numpyro.contrib.control_flow import scan

def mask_inside_plate_inside_scan(y=None, T=10):
    def transition(carry, y_curr):
        x_prev, t = carry
        with numpyro.plate("N", 10, dim=-1):
            with mask(mask=(t < T)):
                x_curr = numpyro.sample('x', dist.Normal(jnp.zeros((10,8)), jnp.ones((10,8))).to_event(1))
                y_curr = numpyro.sample('y', dist.Normal(x_curr, jnp.ones((10,8))).to_event(1), obs=y_curr)
                return (x_curr, t+1), None
    x0 = numpyro.sample('x_0', dist.Normal(jnp.zeros((10,8)), jnp.ones((10,8))).to_event(1))
    print
    x, t = scan(transition, (x0, 0), y, length=T)
    return (x, y)

with numpyro.handlers.seed(rng_seed=0):
    x, y = mask_inside_plate_inside_scan()
```

```
Traceback (most recent call last):
  File "/home/frans/Desktop/test_batch_util_independent.py", line 25, in <module>
    x, y = mask_inside_plate_inside_scan()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/Desktop/test_batch_util_independent.py", line 21, in mask_inside_plate_inside_scan
    x, t = scan(transition, (x0, 0), y, length=T)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/protein_dmm/numpyro/numpyro/contrib/control_flow/scan.py", line 441, in scan
    msg = apply_stack(initial_msg)
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/protein_dmm/numpyro/numpyro/primitives.py", line 53, in apply_stack
    default_process_message(msg)
  File "/home/frans/protein_dmm/numpyro/numpyro/primitives.py", line 28, in default_process_message
    msg["value"] = msg["fn"](*msg["args"], **msg["kwargs"])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/protein_dmm/numpyro/numpyro/contrib/control_flow/scan.py", line 313, in scan_wrapper
    site["fn"] = promote_batch_shape(site["fn"])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/protein_dmm/numpyro/numpyro/distributions/batch_util.py", line 569, in _promote_batch_shape_masked
    new_base_dist = promote_batch_shape(d.base_dist)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/frans/protein_dmm/numpyro/numpyro/distributions/batch_util.py", line 515, in _default_promote_batch_shape
    attr_name = list(d.arg_constraints.keys())[0]
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
``

The problem is that x is a `MaskedDistribution`, whose base_dist is an `Independent` (whose base_dist is `Normal`). Since `Independent` isn't being dispatched to, it gets dispatched to `_default_promote_batch_shape`, which tries to access `arg_constraints`, which aren't there, but under the base_dist.
The error seems to only appear when inside a scan.

I misunderstood your comment before, but I've applied it now.

Reply:
The example is great!! Could you add the above test case to [test_control_flow](https://github.com/pyro-ppl/numpyro/blob/master/test/contrib/test_control_flow.py)? Thanks!

Reply:
I've added it now.